qiskit/transpiler/passes/synthesis/plugin.py

```diff
@@ -0,0 +1,316 @@
+a setuptools entry point in the package metadata. This is done by simply adding
+an ``entry_points`` entry to the ``setuptools.setup`` call in the ``setup.py``
+for the plugin package with the necessary entry points under the
+``qiskit.unitary_synthesis`` namespace. For example::
+
+    entry_points = {
+        'qiskit.unitary_synthesis': [
+            'special = qiskit_plugin_pkg.module.plugin:SpecialUnitarySynthesis',
+        ]
+    },
+
+(note that the entry point ``name = path`` is a single string not a Python
+expression). There isn't a limit to the number of plugins a single package can
+include as long as each plugin has a unique name. So a single package can
+expose multiple plugins if necessary. The name ``default`` is used by Qiskit
+itself and can't be used in a plugin.
+
+Using Plugins
+=============
+
+To use a plugin all you need to do is install the package that includes a
+synthesis plugin. Then Qiskit will automatically discover the installed
+plugins and expose them as valid options for the appropriate
+:func:`~qiskit.compiler.transpile` kwargs and pass constructors. If there are
+any installed plugins which can't be loaded/imported this will be logged to
+Python logging.
+
+To get the installed list of installed unitary synthesis plugins you can use the
+:func:`qiskit.transpiler.passes.synthesis.plugin.unitary_synthesis_plugin_names`
+function.
+
+Plugin API
+==========
+
+Unitary Synthesis Plugins
+-------------------------
+
+.. autosummary::
+   :toctree: ../stubs/
+
+   UnitarySynthesisPlugin
+   UnitarySynthesisPluginManager
+   unitary_synthesis_plugin_names
+
+"""
+
+import abc
+
+import stevedore
+
+
+class UnitarySynthesisPlugin(abc.ABC):
+    """Abstract plugin Synthesis plugin class
+
+    This abstract class defines the interface for unitary synthesis plugins.
+    """
+
+    @property
+    @abc.abstractmethod
+    def supports_basis_gates(self):
+        """Return whether the plugin supports taking ``basis_gates``
+
+        If this returns ``True`` the plugin's ``run()`` method will be
+        passed a ``basis_gates`` kwarg with a list of gate names the target
+        backend supports. For example, ``['sx', 'x', 'cx', 'id', 'rz']``."""
+        pass
+
+    @property
+    @abc.abstractmethod
+    def supports_coupling_map(self):
+        """Return whether the plugin supports taking ``coupling_map``
+
+        If this returns ``True`` the plugin's ``run()`` method will receive
+        two kwargs ``coupling_map`` and ``qubits``. The ``coupling_map`` kwarg
+        will be a :class:`~qiskit.transpiler.CouplingMap` object representing
+        the qubit connectivity of the target backend. The ``qubits`` kwarg will
+        receive a list of integers that represent the qubit indices in the
+        coupling map that unitary is on.
+        """
+        pass
+
+    @property
+    @abc.abstractmethod
+    def supports_approximation_degree(self):
```

Comment:
I was just mirroring the interface we offer on the pass currently this was one of the tunable parameters when initializing the pass. But, yeah I agree it's not the best defined flag, it always seems kind of like a weird knob that didn't give users enough insight into what it was actually doing. I can remove it but it leaves open the question of how to handle this option in the default plugin (since it supports it now)

Reply:
Yeah, I totally see where it comes from — but if we wanted to discourage use of passing an ill-defined value, then API definition is the time to do it...

Reply:
I dropped this from the plugin interface and just updated the transpiler pass to directly set the attribute on the default plugin as a special case for backwards compatibility in: https://github.com/Qiskit/qiskit-terra/pull/6124/commits/cc9fe27fec20bd6257a01a067a6465024bd3ccc6